lib/render.js

```diff
@@ -186,56 +191,52 @@ MapnikSource.prototype._renderMetatile = function(options, callback) {
     // Add reference to the source allowing debug/stat reporting to be compiled.
     options.source = source;
 
-    process.nextTick(function() {
-        // acquire can throw if pool is draining
-        try {
-            source._pool.acquire(function(err, map) {
-                if (err) {
-                    return callback(err);
-                }
-                // Begin at metatile boundary.
-                options.x = meta.x;
-                options.y = meta.y;
-                options.variables.zoom = options.z;
-                map.resize(meta.width, meta.height);
-                map.extent = meta.bbox;
-                try {
-                    source._stats.render++;
-                    var renderStats = {};
-                    var renderStartTime = Date.now();
-                    map.render(image, options, function(err, image) {
-                        process.nextTick(function() {
-                            // Release after the .render() callback returned
-                            // to avoid mapnik errors.
-                            source._pool.release(map);
-                        });
-
-                        if (err) return callback(err);
-                        if (meta.tiles.length > 1) {
-                            renderStats.render = Math.round((Date.now() - renderStartTime) / meta.tiles.length);
-                            sliceMetatile(source, image, options, meta, renderStats, callback);
-                        } else {
-                            renderStats.render = Date.now() - renderStartTime;
-                            encodeSingleTile(source, image, options, meta, renderStats, callback);
-                        }
-                    });
-                } catch(err) {
-                    process.nextTick(function() {
-                        // Release after the .render() callback returned
-                        // to avoid mapnik errors.
-                        source._pool.release(map);
-                    });
-                    return callback(err);
-                }
-            });
-        } catch (err) {
-            return callback(err);
-        }
+    const keys = meta.tiles.map(function(tile) {
+        return options.format + ',' + tile.join(',');
     });
 
+    // acquire can throw if pool is draining
+    try {
+        this._pool.acquire((err, map) => {
+            if (err) return callback(err, keys);
+
+            // Begin at metatile boundary.
+            options.x = meta.x;
+            options.y = meta.y;
+            options.variables.zoom = options.z;
+            map.resize(meta.width, meta.height);
+            map.extent = meta.bbox;
+            try {
+                source._stats.render++;
+                const renderStats = {};
+                const renderStartTime = Date.now();
+                map.render(image, options, (err, image) => {
+                    this._pool.release(map);
```

Comment:
Regarding the comment: `Release after the .render() callback returned	to avoid mapnik errors.` to justify the `.nectTick()`. Why do we avoid mapnik errors now?

Reply:
I don't know why that was there; to my knowledge there is nothing in node-mapnik that requires the map object to live for an extra tick.
`_pool.release(map)` just does `delete map` (plus whatever it needs to do internally in the pool) and, since you are no longer using `map`, it should be perfectly safe. If it wasn't that would mean there is some memory being incorrectly shared by `map` and `err` or `image`. That case would be a (big) bug in node-mapnik and it should be fixed there instead of artificially delay the resource release.

Reply:
Thanks for the explanation.